bench: enable wallet creation benchmarks on all platforms

furszy opened the pull request:
Simple fix for #29816.

Since the wallet is appended to the global `WalletContext` during
creation, merely calling `reset()` on the benchmark shared_pointer
is insufficient to destruct the wallet. This no destruction of the
wallet object results in keeping the db connection open, which
was causes the `fs::remove_all()` failure on Windows.

fanquake commented:
The native Windows CI (https://github.com/bitcoin/bitcoin/actions/runs/9112695910/job/25052498545?pr=30122#step:24:64) is failing with:
```bash
Run src\bench_bitcoin.exe -sanity-check
Running with -sanity-check option, output is being suppressed as benchmark results will be useless.
Error: Process completed with exit code 1.
```

fanquake commented:
It'd be good if the commit message could explain why this is being done. It seems like the current approach is just to try and refactor away the Windows bug, even though it might not have been properly identified (prior to https://github.com/bitcoin/bitcoin/issues/29816#issuecomment-2115302513). However now that the bug might have been found, it'd be good to at least document what it is, and if possible, fix/workaround it in a more targetted way, rather than refactor code that is currently perfectly fine for all other platforms.

furszy commented:
> It'd be good if the commit message could explain why this is being done. It seems like the current approach is just to try and refactor away the Windows bug, even though it might not have been properly identified (prior to [#29816 (comment)](https://github.com/bitcoin/bitcoin/issues/29816#issuecomment-2115302513)). However now that the bug might have been found, it'd be good to at least document what it is, and if possible, fix/workaround it in a more targetted way, rather than refactor code that is currently perfectly fine for all other platforms.

Sure with a small nuance. This code is working fine for all other platforms but that does not mean the code is perfect.
When I wrote this benchmark, I was probably more concerned about the number of files the benchmark could open in parallel than the delays/inaccuracies introduced by the, OS dependent, directory removal line. Essentially, we shouldn't be removing a directory within the benchmarked code scope.
But yeah, np on splitting this in two well described commits.

furszy committed:
bench: bugfix, properly release wallet before erasing directory

Since the wallet is appended to the global WalletContext during
creation, merely calling 'reset()' on the benchmark shared_pointer
is insufficient to destruct the wallet. This no destruction of the
wallet results in the db connection remaining open, which was the
cause of the 'fs::remove_all()' error in Windows.

Co-Authored-By: Hennadii Stepanov <32963518+hebasto@users.noreply.github.com>

hebasto reviewed (approved):
ACK 7323bf6a2ecc3a2f1e2fdebce53b961767b06e08.

Maybe deduplicate code:
```diff
--- a/src/bench/wallet_create.cpp
+++ b/src/bench/wallet_create.cpp
@@ -35,8 +35,8 @@ static void WalletCreate(benchmark::Bench& bench, bool encrypted)
     std::vector<bilingual_str> warnings;
 
     int round = 0;
-    fs::path wallet_path = test_setup->m_path_root / strprintf("test_wallet%d_%d", round, random.rand32()).c_str();
     bench.run([&] {
+        auto wallet_path = test_setup->m_path_root / strprintf("test_wallet%d_%d", round++, random.rand32()).c_str();
         auto wallet = CreateWallet(context, wallet_path.utf8string(), /*load_on_start=*/std::nullopt, options, status, error_string, warnings);
         assert(status == DatabaseStatus::SUCCESS);
         assert(wallet != nullptr);
@@ -44,7 +44,6 @@ static void WalletCreate(benchmark::Bench& bench, bool encrypted)
         // Unload wallet and update wallet directory for the next round
         RemoveWallet(context, wallet, /*load_on_start=*/ std::nullopt);
         UnloadWallet(std::move(wallet));
-        wallet_path = test_setup->m_path_root / strprintf("test_wallet%d_%d", ++round, random.rand32()).c_str();
     });
 }
 
```
?

furszy commented:
Applied, thanks @hebasto 👍🏼.

hebasto reviewed (approved):
re-ACK b6b2d822eb06d705e6ec4318211163e5862008a6.

furszy commented on src/bench/wallet_create.cpp:
```diff
@@ -34,16 +34,16 @@ static void WalletCreate(benchmark::Bench& bench, bool encrypted)
+        // Release wallet
         RemoveWallet(context, wallet, /*load_on_start=*/ std::nullopt);
         UnloadWallet(std::move(wallet));
-        fs::remove_all(wallet_path);
```
> However, there is also a call to RemoveWallet, and UnloadWallet in this WalletCreate benchmark. The additional overhead of the remove_all should be fine? The benefit would be that the number of folders is kept constant and does not pile up as the bench runs, which could (in theory?) lead to issues when running for a long time?

hmm, I don't think they are comparable. The overhead of `remove_all` depend on external factors such us the OS and the hard drive. But agree that running this for a long time isn't good neither. Seems that the best solution would be to create a mechanism to cleanup state in-between benchmark rounds in the framework.
Will drop the second commit for now, np.

maflcko reviewed (approved):
utACK 7e30fdaa8fd7cc572505b7eac1b25f4ebe3b5443

furszy commented on src/bench/wallet_create.cpp:
```diff
@@ -34,16 +34,16 @@ static void WalletCreate(benchmark::Bench& bench, bool encrypted)
-    fs::path wallet_path = test_setup->m_path_root / strprintf("test_wallet_%d", random.rand32()).c_str();
+    int round = 0;
     bench.run([&] {
+        auto wallet_path = test_setup->m_path_root / strprintf("test_wallet%d_%d", round++, random.rand32()).c_str();
```
No needed. I think I had the random path calculated outside of the loop before.
With the current code, neither the counter nor the random is needed. Thanks.

furszy commented on src/bench/wallet_create.cpp:
```diff
@@ -34,29 +34,25 @@ static void WalletCreate(benchmark::Bench& bench, bool encrypted)
-    fs::path wallet_path = test_setup->m_path_root / strprintf("test_wallet_%d", random.rand32()).c_str();
+    int round = 0;
     bench.run([&] {
+        auto wallet_path = test_setup->m_path_root / strprintf("test_wallet%d_%d", round++, random.rand32()).c_str();
```
done

furszy commented on src/bench/wallet_create.cpp:
```diff
@@ -34,29 +34,25 @@ static void WalletCreate(benchmark::Bench& bench, bool encrypted)
     std::vector<bilingual_str> warnings;
 
-    fs::path wallet_path = test_setup->m_path_root / strprintf("test_wallet_%d", random.rand32()).c_str();
+    auto wallet_path = fs::PathToString(test_setup->m_path_root / fs::PathFromString(strprintf("test_wallet")));
```
done. thanks

hebasto reviewed (approved):
re-ACK 7c8abf3c2001152423da06d25f9f4906611685ea, I agree with changes since my recent [review](https://github.com/bitcoin/bitcoin/pull/30122#pullrequestreview-2061694682).

kevkevinpal commented:
utACK [7c8abf3](https://github.com/bitcoin/bitcoin/pull/30122/commits/7c8abf3c2001152423da06d25f9f4906611685ea)
